### ISSUE TITLE
Not Compiling everything on Fedora 36 (GCC 12)

### ISSUE THREAD

**Issue by tmayoff** [objectcomputing/OpenDDS#3495]
I brought this up in a PR but I'm having trouble compiling OpenDDS in Fedora 36
I ran these commands on both my Fedora machine and a Debian 11 VM with a freshly cloned version of OpenDDS.
```
./configure --tests
make -j4
```

On Debian it works fine, and on Fedora it compiles something, but when I try to run the tests it can't find the DCPSInfoRepo in the `bin` folder as the only thing in there is the opendds_idl program. The output from the configure step seems to be the same on both machines, I'm not sure where it's going wrong. I'd like to contribute more to OpenDDS but this is making verifying changes difficult

**Comment by mitza-oci** [comment]
Please attach logs from both the "configure" and "make" steps.  

Another option would be to add a GitHub Actions job for it.  In `.github/workflows/build_and_test.yml` there's an example of a build step that uses a docker container to build in a different distribution/version than what GitHub Actions offers on their runners.

**Comment by tmayoff** [comment]
This is the run with an error I also encountered only on Fedora (it would sometimes error out here and other times continue and just not compile things), https://github.com/tmayoff/OpenDDS/runs/6327318193?check_suite_focus=true

**Comment by mitza-oci** [comment]
Try a newer ACE_TAO by adding the `--ace-github-latest` argument to the configure script.

According to https://fedoraproject.org/wiki/Changes/GNUToolchainF36, Fedora 36 is using a very new GCC so we may need a change to ACE_TAO (and perhaps even OpenDDS) to adapt to it.  So far we've tested OpenDDS with GCC 11 which up until [today](https://gcc.gnu.org/gcc-12) was the latest official release.

**Comment by tmayoff** [comment]
No luck with the --ace-github-latest, and the GCC version seems to track with what I'm seeing I setup a fedora35 run and it seems to be working, is this an issue with ACE then should I open an issue there? I'd be willing to work on fixing

**Comment by mitza-oci** [comment]
A few other things to try first:
* `--doc-group3` instead of `--ace-github-latest` will use ACE version 7.0.6
* There is no configure script option to automate this, but you can try the master branch from https://github.com/DOCGroup/ACE_TAO
  * Before running configure clone https://github.com/DOCGroup/ACE_TAO so it creates the `ACE_TAO` directory under the OpenDDS root
  * Then clone https://github.com/DOCGroup/MPC so it's nested under ACE (in `ACE_TAO/ACE/MPC`)

If neither of those work then it will need to be a new Issue / PR in ACE_TAO

**Comment by tmayoff** [comment]
`--doc-group3` didn't work, I'll have to check the manual setup later

**Comment by tmayoff** [comment]
Manual setup done and still no, however the first time the `make` command exited with no errors logs below.

[configure.log](https://github.com/objectcomputing/OpenDDS/files/8644067/configure.log)
[make.log](https://github.com/objectcomputing/OpenDDS/files/8644068/make.log)
 
If I run make a second time I get the same error as before

**Comment by tmayoff** [comment]
I just realized on fedora I use clang (version 14), although I get the same errors on gcc. I just tried compiling ACE_TAO manually and same errors I'll open an issue over there

**Comment by mitza-oci** [comment]
I didn't realize it was configured for clang.  Please run `/usr/bin/clang++ -v` and paste the output here.

**Comment by tmayoff** [comment]
My personal machine is using clang, however the Github workflows I added should all be using gcc-12
```
clang version 14.0.0 (Fedora 14.0.0-1.fc36)
Target: x86_64-redhat-linux-gnu
Thread model: posix
InstalledDir: /usr/bin
Found candidate GCC installation: /usr/bin/../lib/gcc/x86_64-redhat-linux/12
Selected GCC installation: /usr/bin/../lib/gcc/x86_64-redhat-linux/12
Candidate multilib: .;@m64
Candidate multilib: 32;@m32
Selected multilib: .;@m64
```

**Comment by mitza-oci** [comment]
For another data point, we have a working scoreboard build using this clang (from llvm.org's apt repo):

```
Ubuntu clang version 14.0.1-++20220426083136+0e27d08cdeb3-1~exp1~20220426083221.132
Target: x86_64-pc-linux-gnu
```

**Comment by tmayoff** [comment]
Is there any more information or anything I can provide I'm not sure what needs to be done going forward

**Comment by mitza-oci** [comment]
It does need to be fixed in the ACE_TAO repo before we can proceed with OpenDDS.  I haven't had a lot of time to dig into it, but the GCC change seems to be https://gcc.gnu.org/bugzilla/show_bug.cgi?id=51577 (I'm sure GCC is correct with respect to the C++ spec, but it happens to break what TAO is doing).  It may be good to find out if GCC developers or other open source projects have suggested workarounds for it.

**Comment by Sascha-Clausen** [comment]
To build and use ACE/TAO with GCC 12 you have to add

`#define ACE_ANY_OPS_USE_NAMESPACE`

to your ACE/config.h

https://www.gnu.org/software/gcc/gcc-12/changes.html
Two-stage name lookup for dependent operator expressions has been corrected (PR51577)

https://gcc.gnu.org/bugzilla/show_bug.cgi?id=51577
https://gcc.gnu.org/bugzilla/show_bug.cgi?id=51577#c23
